### PR TITLE
fix: emit will-navigate for sandboxed contents

### LINKED ISSUE #8841: `will-navigate` event not fired when in sandbox mode
Thanks for Electron! Amazing project :)

* Electron version: 1.6.2
* Operating system: macOS Sierra (10.12.3)

It happens when `BrowserWindow` is opened with `sandbox: true`. It might be on purpose, but I couldn't find any Electron/Chromium documentation that explains why. So I'm suspecting it's a bug. I can submit a PR with a failing test to this repo if you agree that it's a bug.

### Expected behavior

`webContents`' `will-navigate` event should fire when user reloads page fx.

### Actual behavior

It does not fire.

### How to reproduce

package.json:

```json
{
  "name": "electron-sandbox-will-navigate-bug",
  "version": "0.0.0",
  "main": "main.js",
  "dependencies": {
    "electron": "^1.6.2"
  }
}
```

main.js:

```js
const {app, BrowserWindow} = require('electron')
const path = require('path')

app.on('ready', function() {
  const win = new BrowserWindow({
    webPreferences: {
      sandbox: true
    }
  })

  const {webContents} = win
  webContents.on('will-navigate', () => {
    console.log('will-navigate')
  })
  webContents.on('did-navigate', () => {
    console.log('did-navigate')
  })

  win.loadURL('file://' + path.join(__dirname, 'index.html'))
})
```

index.html:

```html
<!doctype html>
<html>
  <body>
    <button onclick="window.location.reload()">Reload</button>
  </body>
</html>
```

Perform these steps:

- Run `npm install`
- Run `./node_modules/.bin/electron .`
- Click the Reload button in the browser window
- Observe the Terminal output

Expected output:

```
did-navigate
will-navigate
did-navigate
```

Actual output:

```
did-navigate
did-navigate
```

(will-navigate is not fired).

If you outcomment the line `sandbox:true` in `main.js`, then it works as expected.

### PR DESCRIPTION
**Pull request by nornagon**
#### Description of Change
Fixes #8841.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed "will-navigate" event not being emitted for sandboxed contents.

### CODE REVIEW

**Comment by nornagon** [comment]
NB. this approach was previously attempted here https://github.com/electron/electron/pull/20050, but wasn't accepted because of an assertion that `ShouldFork` causes the renderer process to restart. I've tested this code locally and that doesn't seem to be the case—the navigation does not cause a new process to start. Cross-domain navigations still cause a new renderer, as expected, but within a single domain the process is reused.

Here's a test fiddle: https://gist.github.com/f4dfb4cd13e1a433b79a118a94a98965

Try watching `ps` or equivalent process monitor and watch for changing PIDs.

File: shell/browser/common_web_contents_delegate.cc [hunk 1]
```diff
@@ -292,8 +292,21 @@ content::WebContents* CommonWebContentsDelegate::OpenURLFromTab(
   load_url_params.should_replace_current_entry =
       params.should_replace_current_entry;
   load_url_params.is_renderer_initiated = params.is_renderer_initiated;
+  load_url_params.started_from_context_menu = params.started_from_context_menu;
   load_url_params.initiator_origin = params.initiator_origin;
-  load_url_params.should_clear_history_list = true;
```
Comment (deepak1556) [inline shell/browser/common_web_contents_delegate.cc]: We maintain our own navigation controller, so its necessary to clear the chromium history list.

**Comment by deepak1556** [comment]
To add some notes on this mechanism:

`ShouldFork` by itself doesn't cause process restart, it gives access to us for all render initiated navigations and lets us decide whether we want it to turn into a browser process initiated navigation so that we can force our process policy through `ShouldOverrideSiteInstanceForNavigation`. There used to be bug with sandbox renderers to use this mechanism but it seems to have improved.

/cc @zcbenz incase I missed a scenario.

File: spec-main/api-browser-window-spec.ts [hunk 2]
```diff
@@ -434,110 +434,128 @@ describe('BrowserWindow module', () => {
...
-
-    describe('will-navigate event', () => {
-      it('allows the window to be closed from the event listener', (done) => {
-        w.webContents.once('will-navigate', () => {
-          w.close()
-          done()
-        })
-        w.loadFile(path.join(fixtures, 'pages', 'will-navigate.html'))
+  for (const sandbox of [false, true]) {
+    describe(`navigation events${sandbox ? ' with sandbox' : ''}`, () => {
```
Comment (deepak1556) [inline spec-main/api-browser-window-spec.ts]: Can you add a test for `will-navigate` in the case when navigating from non-file scheme to file scheme ?

File: shell/browser/common_web_contents_delegate.cc (hunk 1 above)
Comment (nornagon) [reply shell/browser/common_web_contents_delegate.cc]: The tests aren't failing. How would I detect that this is causing incorrect behavior?

Comment (nornagon) [reply shell/browser/common_web_contents_delegate.cc]: FWIW removing this fixed a `DCHECK(!replace)` in the navigation controller.

File: spec-main/api-browser-window-spec.ts (hunk 2 above)
Comment (nornagon) [reply spec-main/api-browser-window-spec.ts]: It seems that results in `"Not allowed to load local resource: file://..."`, in both the sandbox and non-sandbox situations, with this patch.

File: shell/browser/common_web_contents_delegate.cc (hunk 1 above)
Comment (deepak1556) [reply shell/browser/common_web_contents_delegate.cc]: Hmm the non-sandboxed renderer history are unaffected by this change, tested with:

* load `file://index.html`
* window.history.length == 1
* window.location.reload() // creates new process
* window.location = "https://example.com" // creates new process
* window.location.reload() // creates new process
* window.history.length == 4 but should be 2 , this bug is not related to this change.

So seems like custom navigation controller still takes control for non-sandboxed renderers.

Comment (nornagon) [reply shell/browser/common_web_contents_delegate.cc]: Sounds like we're all OK on this front :)

**Pull request merged**